passes/source2il.nim

```diff
@@ -30,20 +31,117 @@ type
 
 using
   c: var Context
+  t: InTree
   bu: var Builder[NodeKind]
 
 template addType(c; kind: NodeKind, body: untyped): uint32 =
   c.types.subTree kind:
     body
   (let r = c.numTypes; inc c.numTypes; r.uint32)
 
+proc typeToIL(c; typ: TypeKind): uint32 =
+  case typ
+  of tkBool:
+    c.addType Int: c.types.add(Node(kind: Immediate, val: 1))
+  of tkInt, tkError:
+    c.addType Int: c.types.add(Node(kind: Immediate, val: 8))
+  of tkFloat:
+    c.addType Float: c.types.add(Node(kind: Immediate, val: 8))
+
 proc exprToIL(c; t: InTree, n: NodeIndex, bu): TypeKind
 
 proc exprToIL(c; t: InTree, n: NodeIndex): (NodeSeq, TypeKind) =
   var bu = initBuilder[NodeKind]()
   result[1] = exprToIL(c, t, n, bu)
   result[0] = finish(bu)
 
+proc binaryArithToIL(c; t; n: NodeIndex, name: string, bu): TypeKind =
+  ## Analyzes and emits the IL for a binary arithmetic operation.
+  if t.len(n) != 3:
+    bu.add Node(kind: IntVal, val: 0) # add at least something
+    return tkError
+
+  let
+    (_, a, b)  = t.triplet(n)
+    (eA, typA) = exprToIL(c, t, a)
+    (eB, typB) = exprToIL(c, t, b)
+
+  let op =
+    case name
+    of "+": Add
+    of "-": Sub
+    else:   unreachable()
+
+  if typA == typB and typA in {tkInt, tkFloat}:
+    let typ = c.typeToIL(typA)
+
+    bu.subTree op:
+      bu.add Node(kind: Type, val: typ)
+      bu.add eA
+      bu.add eB
+
+    result = typA
+  else:
+    bu.add Node(kind: IntVal)
+    result = tkError
+
+proc relToIL(c; t; n: NodeIndex, name: string, bu): TypeKind =
+  ## Analyzes and emits the IL for a relational operation.
+  if t.len(n) != 3:
+    bu.add Node(kind: IntVal)
+    return tkError
+
+  let
+    (_, a, b)  = t.triplet(n)
+    (eA, typA) = exprToIL(c, t, a)
+    (eB, typB) = exprToIL(c, t, b)
+
+  let (op, valid) =
+    case name
+    of "==": (Eq, {tkBool, tkInt, tkFloat})
+    of "<":  (Lt, {tkInt, tkFloat})
+    of "<=": (Le, {tkInt, tkFloat})
+    else:   unreachable()
+
+  if typA == typB and typA in valid:
+    bu.subTree op:
+      bu.add Node(kind: Type, val: c.typeToIL(typA))
+      bu.add eA
+      bu.add eB
+
+    result = tkBool
+  else:
+    bu.add Node(kind: IntVal)
+    result = tkError
+
+proc notToIL(c; t; n: NodeIndex, bu): TypeKind =
+  if t.len(n) != 2:
+    bu.add Node(kind: IntVal)
+    return tkError
```

Comment:
Might be worth templatizing this code similar to `checkMinSons` either now or later.